src/telemetry/telemetryReporter.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import Dap from '../dap/api';
+import { HighResolutionTime, calculateElapsedTime } from '../utils/performance';
+import { OpsReportBatcher, UnbatchedOpReport } from './opsReportBatch';
+import { EventEmitter } from '../common/events';
+
+export type TelemetryEntityProperties = object;
+export type OutcomeAndTime = { time: number, succesful: boolean };
+
+enum RequestOutcome {
+  Succesful,
+  Failed
+}
+
+export class DapRequestTelemetryReporter {
+  private readonly _rawTelemetryReporter: RawTelemetryReporterToDap;
+
+  public constructor(dap: Dap.Api) {
+    this._rawTelemetryReporter = new RawTelemetryReporterToDap(dap);
+  }
+
+  reportErrorWhileHandlingDapMessage(dapCommand: string, receivedTime: HighResolutionTime, error: unknown) {
+    this.reportWithElapsedTime(receivedTime, dapCommand, extractErrorDetails(error), RequestOutcome.Failed);
+  }
+
+  reportSuccesfullyHandledDapMessage(dapCommand: string, receivedTime: HighResolutionTime) {
+    this.reportWithElapsedTime(receivedTime, dapCommand, {}, RequestOutcome.Succesful);
+  }
+
+  private reportWithElapsedTime(receivedTime: [number, number], dapCommand: string, properties: TelemetryEntityProperties, outcome: RequestOutcome) {
+    const elapsedTime = calculateElapsedTime(receivedTime);
+    this._rawTelemetryReporter.report('dap/' + dapCommand, { ...properties, time: elapsedTime, succesful: outcome === RequestOutcome.Succesful });
+  }
+}
+
+export class CdpTelemetryReporter {
+  private readonly _batcher = new OpsReportBatcher();
+
+  public constructor(private readonly _rawTelemetryReporter: RawTelemetryReporter) {
+    this._rawTelemetryReporter.flush.event(() => {
+      this._rawTelemetryReporter.report('cdpOperations', this._batcher.batched());
+    });
+  }
+
+  reportErrorWhileHandlingEvent(cdpEventName: string, receivedTime: HighResolutionTime, error: unknown) {
+    this.reportWithElapsedTime(receivedTime, cdpEventName, extractErrorDetails(error), RequestOutcome.Failed);
+  }
+
+  reportSuccesfullyHandledEvent(cdpEventName: string, receivedTime: HighResolutionTime) {
+    this.reportWithElapsedTime(receivedTime, cdpEventName, {}, RequestOutcome.Succesful);
+  }
+
+  private reportWithElapsedTime(receivedTime: HighResolutionTime, cdpEventName: string, properties: TelemetryEntityProperties, outcome: RequestOutcome) {
+    // Floating point numbers use too much space and we don't need that much precision, we use integers to use less space when batching timings
+    const elapsedTime = Math.ceil(calculateElapsedTime(receivedTime));
```

Comment:
Slightly different from the elapsedTime from the DAP reporter? Should we just do the Math.ceil inside `calculateElapsedTime`?

Reply:
Space is not an issue on the DAP reporter, so I'm inclined to only do it for CDP.